functions/go/set-namespace/transformer/namespace.go

```diff
@@ -18,206 +18,310 @@ import (
 	"strings"
 
 	"github.com/GoogleContainerTools/kpt-functions-sdk/go/fn"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-func SetNamespace(rl *fn.ResourceList) (bool, error) {
-	tc := NamespaceTransformer{}
+// Run provides the main workflow to update the ResourceList.Items "namespace" value.
+func Run(rl *fn.ResourceList) (bool, error) {
+	tc := SetNamespace{}
 	// Get "namespace" arguments from FunctionConfig
 	err := tc.Config(rl.FunctionConfig)
 	if err != nil {
-		rl.Results = append(rl.Results, fn.ErrorResult(err))
-		return false, nil
+		rl.Results = append(rl.Results, fn.ErrorConfigObjectResult(err, rl.FunctionConfig))
+		return true, nil
 	}
 	// Update "namespace" to the proper resources.
-	oldNamespaces := map[string]struct{}{}
-	count := tc.Transform(rl.Items, oldNamespaces)
-	result := tc.LogResults(rl, count, oldNamespaces)
-	rl.Results = append(rl.Results, result)
+	results := tc.Transform(rl.Items)
+	rl.Results = append(rl.Results, results...)
 	return true, nil
 }
 
-func (p *NamespaceTransformer) LogResults(rl *fn.ResourceList, count int, oldNamespaces map[string]struct{}) *fn.Result {
-	if len(p.Errors) != 0 {
-		errMsg := strings.Join(p.Errors, "\n")
-		return fn.GeneralResult(errMsg, fn.Error)
-	}
-	if dupErr := fn.CheckResourceDuplication(rl); dupErr != nil {
-		return fn.ErrorResult(dupErr)
-	}
-	if count == 0 {
-		msg := fmt.Sprintf("all namespaces are already %q. no value changed", p.NewNamespace)
-		return fn.GeneralResult(msg, fn.Info)
-	}
-
-	oldNss := []string{}
-	for oldNs := range oldNamespaces {
-		oldNss = append(oldNss, `"`+oldNs+`"`)
-	}
-	msg := fmt.Sprintf("namespace %v updated to %q, %d value(s) changed", strings.Join(oldNss, ","), p.NewNamespace, count)
-	return fn.GeneralResult(msg, fn.Info)
+// SetNamespace defines structs to parse KRM resource "SetNamespace" (the custom function config) and "ConfigMap" data.
+// it provides the method "Config" to read the function configs from ResourceList.FunctionConfig
+// it provides the method "Transform" to change the "namespace" and update the "config.kubernetes.io/depends-on" annotation.
+type SetNamespace struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	NewNamespace      string            `json:"namespace,omitempty"`
+	NamespaceMatcher  string            `json:"namespaceMatcher,omitempty"`
+	Data              map[string]string `json:"data,omitempty"`
```

Comment:
IMO it makes more sense to have a struct that just defines the `SetNamespace` type and not conflate it with `ConfigMap`. This is a little bit confusing, and I think it would be more clear what is happening here if we separate another struct called `ConfigMap` for the data field (or use the `corev1.ConfigMap` struct).

Reply:
Thanks! That works for me.